various small fixes

0pcom opened the pull request:
Changes:	
-	fix hidden flags nil pointer error on mac & windows ( --all flag error)
-	optimized `run-source` & `run-vpnsrv` makefile directives and related subdirectives
-	fix visor read config from stdin for systray

On that last one, runVisor now can accept a config passed to it as input (or nil). Running with systray was trying to read the config a second time, calling for initConfig . When the config comes via stdin there is no second read possible.

Additions:
-	`prepare` and `prepare-systray` makefile directives
-	`run-systray` makefile directive
-	`run-systray-test` makefile directive


How to test this PR:
```
make run-systray-test
```

Commit:
various small fixes

Commit:
make format check

Commit:
run-systray makefile directive

Commit:
disable systray when visor is run as root - browser cannot launch as root

0pcom commented:
In tests I discovered that the systray app is not appropriate to run as root on linux and does not function as intended. It possibly could but it cannot directly launch the browser as root. I have disabled systray if the root user is detected on macOS or linux.

privilege elevation for the vpn client / server is still yet to be addressed.

Commit:
require root user to run VPN client & server

ersonp commented on local/.gitignore:
```diff
@@ -0,0 +1,4 @@
+# Ignore everything in this directory
```
We are already ignoring the contents [here](https://github.com/the-skycoin-project/skywire/blob/b158522464cfabdf176898b8454628c5b5a8a34d/.gitignore#L16) and I  also do not see the reason why we need to push an empty folder to the repo. Since after starting the visor with sudo we need to delete the folder again anyway to start a visor without sudo.

0pcom replied:
This was something that @jdknives mentioned doing last we spoke. To address the permissions errors between runs as root and not. The issue still remains, with all the files in the folder though not the folder itself. so perhaps that is why it wasn't done. This is an issue I am attempting to address more thoroughly. Doesn't work as intended - will remove.

0pcom replied:
removed

ersonp commented:
> In tests I discovered that the systray app is not appropriate to run as root on linux and does not function as intended. It possibly could but it cannot directly launch the browser as root. I have disabled systray if the root user is detected on macOS or linux.
> 
> privilege elevation for the vpn client / server is still yet to be addressed.

we need to set `cap_net_admin` by `sudo setcap 'cap_net_admin+p' ./apps/vpn-client` and `sudo setcap 'cap_net_admin+p' ./apps/vpn-server`

andi think we should provide a `WARN` log in case if the systray app is started as root.

mrpalide commented on cmd/apps/skysocks-client/skysocks-client.go:
```diff
@@ -28,9 +30,78 @@ const (
+
+var rootCmd = &cobra.Command{
+	Use:   "skysocks",
+	Short: "Skywire SOCKS5 Proxy Server",
```
Seems short description is wrong here

0pcom replied:
changed

mrpalide commented on cmd/apps/skychat/skychat.go:
```diff
@@ -34,62 +35,95 @@ const (
-var embededFiles embed.FS
+var rootCmd = &cobra.Command{
+	Use:   "skysocks",
+	Short: "Skywire SOCKS5 Proxy Server",
```
Should be skychat, not skysocks.

0pcom replied:
changed

mrpalide commented on cmd/apps/vpn-client/vpn-client.go:
```diff
@@ -19,142 +21,193 @@ import (
-	if err := serverPK.UnmarshalText([]byte(*serverPKStr)); err != nil {
-		fmt.Printf("Invalid VPN server pub key: %v\n", err)
+	if (thisUser.Username != "root") && ((skyenv.OS == "linux") || (skyenv.OS == "mac")) {
+		fmt.Println("vpn client must be run as root")
```
No, not root.
We handle administrator privilege in Windows Installer, and in Mac or Linux, need set `net_cap_admin` value for `vpn-client` binary.

0pcom replied:
reverted to original state

mrpalide commented on cmd/apps/vpn-client/vpn-client.go:
```diff
@@ -19,142 +21,193 @@ import (
+		if serverPKStr == "" {
+			// TODO(darkrengarius): fix args passage for Windows
+			//*serverPKStr = "03e9019b3caa021dbee1c23e6295c6034ab4623aec50802fcfdd19764568e2958d"
+			fmt.Println("VPN server pub key is missing")
```
Actually never read key from config and told `VPN server pub key is missing`, even when I set it.
```
[2022-04-12T12:58:01+04:30] INFO (STDOUT) [proc:vpn-client:db6f0128f6da48b49f298d6b4d0df125]: VPN server pub key is missing
```

0pcom replied:
check it again, I wasn't able to reproduce this after my changes

0pcom replied:
i've reverted the apps back to how they are on develop, which should address this. I did reproduce it at one point but seems ok now.

mrpalide commented on cmd/apps/skychat/skychat.go:
```diff
@@ -34,62 +35,95 @@ const (
-func main() {
-	appC = app.NewClient(nil)
-	defer appC.Close()
+		fmt.Print("Successfully started skychat.")
```
Skychat never run, and get this confusing log
```
[2022-04-12T12:53:23+04:30] INFO (STDOUT) [proc:skychat:140eb2010c7d43128ce42e3e7be6af34]: Successfully started skychat.Serving HTTP onddrlisten tcp: address ddr: missing port in address
```
Seems cannot read from config, same as vpn-client.

0pcom replied:
this was related to the short command being wrong, per your previous comment

ersonp commented:
One small change is needed for skywire windows. In windows `config gen` the only app that needs to be included is `vpn-client` as all other apps are not supported.

Commit:
disable browser-launching systray buttons as root ; disable apps that are not found at path specified in config during config gen with small exceptions

Commit:
disable browser-launching systray buttons as root ; disable apps that are not found at path specified in config during config gen with small exceptions

0pcom commented:
On the systray app I changed the behavior from being disabled as root to only hiding the buttons that launch the browser when the process is run as root.

 on the config gen @ersonp I changed it to detect if the applications are at the bin path, and if not they are added to the list of apps to disabled
 
 exceptions to this behavior are in two instances;

* the config is being written to stdout (`-n` flag)
* the cli command was started with `go run`

most of the other stuff mentioned has been cleaned up

Commit:
fix app autodisable in config gen

Commit:
fail on incorrect or insufficient write permissions - prevent writing as root to non root-owned path and fail before attempted non-root write to root-owned path for skywire-cli config gen and skywire-visor

Commit:
fail on incorrect or insufficient write permissions - prevent writing as root to non root-owned path and fail before attempted non-root write to root-owned path for skywire-cli config gen and skywire-visor

0pcom commented:
I've prevented writing to non-root owned path as root and vice versa. With skywire-cli config gen and with skywire-visor

@ersonp I had to revert the `-o` flag for config gen to the way I had it before your recent PR.

At issue is flags.Changed not working right.. my workaround was to have the output be actually unset and set the value after it's detected that it contains an empty string, as well as to record that it was unset into a boolean to be used for overriding the output of the `-p` flag default config path

I was otherwise unable to get the `-o` flag to override the `-p` flag for config gen.

Tomorrow I need to add a flag which will generate one default config at a set location in the user space, which references the users.db and local folder along the same path as the config.

Commit:
implement -u --user flag for skywire-cli config gen & skywire-visor

Commit:
fix -u flag imlementation for skywire-visor

Commit:
implement -u --user flag for skywire-cli config gen & skywire-visor

Commit:
implement -u --user flag for skywire-cli config gen & skywire-visor

Commit:
fix default visor config name when unspecified

Commit:
revert apps back to using pflags

ersonp commented:
Can we just limit the scope of this PR to flags and reading from STDIN because the changes with the path is not actually solving the problem but just shifting it. I think it would be better if we fixed it via the read and write functions of all the files in the `/local` folder. If I remember correctly this issue was solved previously and only reappeared a little while ago (as far as I can tell). So I think it would be better if we took a deeper look at it in a different PR.

Commit:
change fail to warn about writing with elevated permissions. Make logging more descriptive around these errors. Add dtection of unpublised changes since commit in logging

0pcom commented:
@ersonp I've changed it to warn about writing with elevated permissions here:

![image](https://user-images.githubusercontent.com/36607567/163420102-8d57dcfa-542f-4c22-8c58-12a4538d1fc1.png)

I want to maintain the failure in the reverse instance as not failing means it runs but doesn't work, and just repeats "waiting for RPC get ready"

![image](https://user-images.githubusercontent.com/36607567/163423699-3d19e90a-f110-4add-84ad-27776176e61d.png)

jdknives commented on cmd/skywire-cli/commands/config/gen.go:
```diff
@@ -261,6 +335,32 @@ var genConfigCmd = &cobra.Command{
 		}
 		//don't write file with stdout
 		if !stdout {
+			thisUser, err := user.Current()
```
Something like `userLvl` or `owner` etc sounds better.

jdknives replied:
Something like `userLvl` or `owner` etc sounds better.

jdknives commented on cmd/skywire-cli/commands/config/gen.go:
```diff
@@ -261,6 +335,32 @@ var genConfigCmd = &cobra.Command{
 		if !stdout {
+			thisUser, err := user.Current()
+			if err != nil {
+				panic(err)
```
What errors could this return? Are they clear enough in themselves? Should we exit cleanly and log something here instead?

jdknives commented on cmd/skywire-cli/commands/config/gen.go:
```diff
@@ -261,6 +335,32 @@ var genConfigCmd = &cobra.Command{
 		if !stdout {
+			thisUser, err := user.Current()
+			if err != nil {
+				panic(err)
```
What errors could this return? Are they clear enough in themselves? Should we exit cleanly and log something here instead?

0pcom replied:
no need to really error here & Ive changed it to fail transparently. This is only checking for root and it will just not change the default from false to true.

jdknives commented on internal/gui/gui.go:
```diff
@@ -64,24 +64,36 @@ var (
-		initQuitBtn()
-
-		go handleUserInteraction(conf, doneCh)
+	if checkRoot() {
```
`isRoot` maybe clearer?

0pcom replied:
done

jdknives commented on internal/gui/gui.go:
```diff
@@ -417,6 +438,24 @@ func handleUserInteraction(conf *visorconfig.V1, doneCh chan<- bool) {
+func handleRootInteraction(conf *visorconfig.V1, doneCh chan<- bool) {
+	for {
+		select {
+		//		case <-mOpenHypervisor.ClickedCh:
```
What are these commented cases doing here?

jdknives commented on internal/gui/gui_unix.go:
```diff
@@ -3,8 +3,23 @@
 }
+
+func checkRoot() bool {
+	thisUser, err := user.Current()
```
Same suggestion as above. Using `this` sounds odd to me.

0pcom replied:
done

jdknives commented on internal/gui/gui_unix.go:
```diff
@@ -3,8 +3,23 @@
+func checkRoot() bool {
+	thisUser, err := user.Current()
+	if err != nil {
+		panic(err)
```
I would prefer to not panic here and log a fatal err with some context.

0pcom replied:
the logging overhead is unnecessary and it doesn't really need to fail, only to return true if root is detected.

this check is for limiting the available  buttons in the systray to ones that are appropriate to launch by the current user. The browser won't be allowed to run as root in the userspace by the system, typically.

jdknives commented on pkg/skyenv/values.go:
```diff
@@ -169,3 +169,9 @@ func Version() string {
 }
+
+// HomePath gets the current user's home folder
+func HomePath() string {
```
What if this returns an error? Will the visor start? If so, where will it start?

0pcom replied:
You are correct, this could cause an odd configuration to be produced if HomePath returns an empty string.

I've now implemented a check on the config gen command to not offer the `-u --user` flag if HomePath returns an empty string. Similarly to how the visor determines if the -u flag should be an option based on the presence of the config or not.

if HomePath returned empty before these changes it _would have been_ obvious from the config gen help menu as it's displayed in the flag description.

jdknives commented on pkg/skyenv/values_darwin.go:
```diff
@@ -22,3 +19,13 @@ func PackageConfig() PkgConfig {
+	var usrconfig PkgConfig
+	usrconfig.Launcher.BinPath = "/Library/Application Support/Skywire/apps"
+	usrconfig.LocalPath = HomePath() + "/.skywire/local"
+	usrconfig.Hypervisor.DbPath = HomePath() + "/.skywire/users.db" //permissions errors if the process is not run as root.
```
What are you trying to say with the comment? It seems it was copied from previous implementation for `opt`. Does it indeed fail if not run as root in HOME?

0pcom replied:
accidental inclusion, and the original comment issue is now basically resolved

jdknives commented on pkg/skyenv/values_windows.go:
```diff
@@ -21,3 +19,13 @@ func PackageConfig() PkgConfig {
+	var usrconfig PkgConfig
+	usrconfig.Launcher.BinPath = "C:/Program Files/Skywire/apps"
+	usrconfig.LocalPath = HomePath() + "/.skywire/local"
+	usrconfig.Hypervisor.DbPath = HomePath() + "/.skywire/users.db" //permissions errors if the process is not run as root.
```
Same question about the validity of the comment as above.

0pcom replied:
accidental inclusion, and the original comment issue is now basically resolved

Commit:
cleanup per @jdknives suggestions

Commit:
make format check
